### PR TITLE
S3 backing for profefe

### PR DESCRIPTION
**Pull request by goller**
Hey there, here is a PR to store profiles in S3.

I've encoded the metadata about the profile within the S3 key like this:

```
/service/profile_type/created_at_unix_time/label1=value1,label2=value2/id
```

Additionally, I tweaked the CLI parameters so that one must specify either badger or s3.

What do you think?

### CODE REVIEW

**Comment by gianarb** [comment]
Just as a bonus point, this code is running across 4 different environments (one each) and we can't complain! :tada:

**Comment by narqo** [comment]
Thank you for the PR; I will take a deeper look a bit later.

Because my experience with s3 is quite limited, could you help to better understand the PR:

1. How do you pass credentials for a bucket to s3 API?

2. As you've mentioned, you store labels as part of the object's key. Do you know if using [s3's User-Defined Object Metadata](https://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#UserMetadata) has any downsides (sorry, I haven't looked deep into the implementation of `storage/s3`'s querying part yet)?

**Comment by goller** [comment]
Hi @narqo !

> 1. How do you pass credentials for a bucket to s3 API?

There are three typical ways of doing this:

a. via IAM (giving permissions to an entire machine)
b. within your dotfiles (.aws/credentials)
c. using environment variables

here are some more specifics:
https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials

Those 3 methods are used for most (all?) of AWS services.

> 2. As you've mentioned, you store labels as part of the object's key. Do you know if using s3's User-Defined Object Metadata has any downsides (sorry, I haven't looked deep into the implementation of storage/s3's querying part yet)?

I had high hopes for the object metadata, but, in the end I was not able to use it.  To get metadata back from s3 one must know the _exact_ key for a single s3 object.  The profefe search filters are time-range and label based, so, in order to use meta data one would need to download the metadata for all profiles stored in s3.

When asked about using object metadata for search, AWS guidance was to use dynamodb instead.

However, S3 does have some options for their listing. Using special key paths I did not need to use another service like dynamo.

I use them here:

https://github.com/profefe/profefe/pull/52/files#diff-8958ece5da1c4fbf5c67668ac7eab2bfR202-R207

```go
	input := &s3.ListObjectsV2Input{
		Bucket:     &s.S3Bucket,
		Prefix:     aws.String(prefix(params)),
		StartAfter: aws.String(startAfter(params)),
		MaxKeys:    aws.Int64(1000),
	}
```

The `Prefix` searches through the s3 objects for just those with keys that start with `Prefix`.  I used the service and profile type to as the prefix.

`StartAfter` positions the list cursor after a key in a list of lexicographical sorted keys with prefix.  I use the created at min timestamp as the StartAfter so all profiles returned are after that time.

File: pkg/storage/s3/s3.go [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+	}
+}
+
+func (s *Store) newDownloader() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.downloader == nil {
+		s.downloader = s3manager.NewDownloaderWithClient(s.svc)
+	}
+}
```
Comment (narqo) [inline pkg/storage/s3/s3.go]: Is there any need in lazy initialisation of `s.uploader` and `s.downloader`? They don't seem ever change and could be instantiated in `NewStore` making the code a bit simplier to follow, doesn't it?

File: pkg/storage/s3/s3.go [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+	limit := params.Limit
+	if limit == 0 {
+		limit = 100
+	}
+
+	input := &s3.ListObjectsV2Input{
+		Bucket:     &s.S3Bucket,
+		Prefix:     aws.String(prefix(params)),
+		StartAfter: aws.String(startAfter(params)),
+		MaxKeys:    aws.Int64(1000),
```
Comment (narqo) [inline pkg/storage/s3/s3.go]: Did you mean to use `limit` variable here?

**Comment by narqo** [comment]
I actually start thinking that if we moved profile id management from `pkg/profefe` to storage, the s3 implementation might be simplier. That is, we could say, that for `storage/s3` the id of the profile is the whole key `/service/type/created/labels`. With that, we didn't need to store separate "dummy" encoded meta object to make both query and get requests.

**Pull request closed without merging**